File: app/src/main/java/com/timotiusoktorio/pencake/ui/orderdetail/OrderDetailActivity.java
package com.timotiusoktorio.pencake.ui.orderdetail;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.MenuItem;
import android.view.View;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.timotiusoktorio.pencake.R;
import com.timotiusoktorio.pencake.data.model.Coordinate;
import com.timotiusoktorio.pencake.data.model.Order;
import com.timotiusoktorio.pencake.data.model.StoreInfo;
import com.timotiusoktorio.pencake.databinding.ActivityOrderDetailBinding;
import com.timotiusoktorio.pencake.ui.BaseActivity;
import com.timotiusoktorio.pencake.ui.adapters.CartRvAdapter;

import javax.inject.Inject;

public class OrderDetailActivity extends BaseActivity implements OnMapReadyCallback {

    private static final String EXTRA_ORDER_JSON = "EXTRA_ORDER_JSON";

    @Inject OrderDetailViewModelFactory vmFactory;

    private ActivityOrderDetailBinding binding;
    private OrderDetailViewModel viewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_order_detail);
        getComponent().inject(this);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        viewModel = ViewModelProviders.of(this, vmFactory).get(OrderDetailViewModel.class);
        viewModel.storeInfoLiveData.observe(this, new Observer<StoreInfo>() {
            @Override
            public void onChanged(@Nullable StoreInfo storeInfo) {
                updateStoreInfo(storeInfo);
            }
        });

        String orderJson = getIntent().getStringExtra(EXTRA_ORDER_JSON);
        if (orderJson == null) {
            throw new IllegalStateException("Order JSON was not sent here as intent extra");
        }

        final Order order = Order.fromJson(orderJson);
        binding.setOrder(order);

        CartRvAdapter adapter = new CartRvAdapter(order.getCart());
        binding.cartRv.setAdapter(adapter);
        binding.cartRv.setLayoutManager(new LinearLayoutManager(this));
        binding.cartRv.addItemDecoration(new DividerItemDecoration(this, LinearLayoutManager.VERTICAL));
        binding.cartRv.setHasFixedSize(true);

        int orderStatusTitleRes = order.getOrderStatus().getTitleRes();
        binding.orderStatusTv.setText(orderStatusTitleRes);
        binding.contactUsButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendEmailIntent(order.getId());
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        final StoreInfo storeInfo = viewModel.storeInfoLiveData.getValue();
        if (storeInfo != null) {
            final Coordinate coordinate = storeInfo.getCoordinate();
            LatLng latLng = new LatLng(coordinate.getLatitude(), coordinate.getLongitude());
            googleMap.addMarker(new MarkerOptions().position(latLng));
            googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 16));
            googleMap.getUiSettings().setAllGesturesEnabled(false);
            googleMap.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
                @Override
                public void onMapClick(LatLng latLng) {
                    Coordinate coordinate = storeInfo.getCoordinate();
                    Uri gmmIntentUri = Uri.parse("google.navigation:q=" + coordinate.getLatitude() + "," + coordinate.getLongitude());
                    Intent mapIntent = new Intent(Intent.ACTION_VIEW, gmmIntentUri);
                    mapIntent.setPackage("com.google.android.apps.maps");
                    if (mapIntent.resolveActivity(getPackageManager()) != null) {
                        startActivity(mapIntent);
                    }
                }
            });
        }
    }

    private void updateStoreInfo(@Nullable final StoreInfo storeInfo) {
        if (storeInfo != null) {
            binding.setStoreInfo(storeInfo);
            binding.directionsIcon.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Coordinate coordinate = storeInfo.getCoordinate();
                    Uri gmmIntentUri = Uri.parse("geo:0,0?q=" +
                            coordinate.getLatitude() + "," +
                            coordinate.getLongitude() + "( " + storeInfo.getName() + ")");

                    Intent mapIntent = new Intent(Intent.ACTION_VIEW, gmmIntentUri);
                    mapIntent.setPackage("com.google.android.apps.maps");
                    if (mapIntent.resolveActivity(getPackageManager()) != null) {
                        startActivity(mapIntent);
                    }
                }
            });
            MapFragment mapFragment = (MapFragment) getFragmentManager().findFragmentById(R.id.mapFragment);
            mapFragment.getMapAsync(this);
        }
    }

    private void sendEmailIntent(String orderId) {
        Intent intent = new Intent(Intent.ACTION_SENDTO);
        intent.setData(Uri.parse("mailto:"));
        intent.putExtra(Intent.EXTRA_EMAIL, new String[] {"support@pencake.com"});
        intent.putExtra(Intent.EXTRA_SUBJECT, "Order Support (" + orderId + ")");
        if (intent.resolveActivity(getPackageManager()) != null) {
            startActivity(intent);
        }
    }

    public static Intent newIntent(FragmentActivity activity, String orderJson) {
        Intent intent = new Intent(activity, OrderDetailActivity.class);
        intent.putExtra(EXTRA_ORDER_JSON, orderJson);
        return intent;
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/orderdetail/OrderDetailActivity.kt
package com.timotiusoktorio.pencake.ui.orderdetail

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.MenuItem
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapFragment
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.timotiusoktorio.pencake.R
import com.timotiusoktorio.pencake.data.model.Order
import com.timotiusoktorio.pencake.data.model.StoreInfo
import com.timotiusoktorio.pencake.data.source.DataManager
import com.timotiusoktorio.pencake.extensions.consume
import com.timotiusoktorio.pencake.extensions.observe
import com.timotiusoktorio.pencake.extensions.toLongDateFormat
import com.timotiusoktorio.pencake.extensions.withViewModel
import com.timotiusoktorio.pencake.ui.BaseActivity
import com.timotiusoktorio.pencake.ui.adapters.CartRvAdapter
import kotlinx.android.synthetic.main.activity_order_detail.*
import java.util.*
import javax.inject.Inject

class OrderDetailActivity : BaseActivity(), OnMapReadyCallback {

    @Inject lateinit var dataManager: DataManager

    private lateinit var viewModel: OrderDetailViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_order_detail)
        component.inject(this)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val order = Order.fromJson(intent.getStringExtra(EXTRA_ORDER_JSON))
                ?: throw IllegalStateException("Order JSON was not sent here as intent extra")

        initViews(order)
        initViewModel()
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean = when (item!!.itemId) {
        android.R.id.home -> consume { finish() }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        viewModel.storeInfoLiveData.value?.let { storeInfo ->
            val coordinate = storeInfo.coordinate
            val latLng = LatLng(coordinate.latitude, coordinate.longitude)
            googleMap.addMarker(MarkerOptions().position(latLng))
            googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 16F))
            googleMap.uiSettings.setAllGesturesEnabled(false)
            googleMap.setOnMapClickListener {
                val uri = Uri.parse("geo:0,0?q=${coordinate.latitude},${coordinate.longitude}(${storeInfo.name})")
                val mapIntent = Intent(Intent.ACTION_VIEW, uri)
                mapIntent.`package` = "com.google.android.apps.maps"
                if (mapIntent.resolveActivity(packageManager) != null) {
                    startActivity(mapIntent)
                }
            }
        }
    }

    private fun initViews(order: Order) {
        val rvAdapter = CartRvAdapter(order.cart.toMutableList()) {}
        cartRv.adapter = rvAdapter
        cartRv.layoutManager = LinearLayoutManager(this)
        cartRv.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
        cartRv.setHasFixedSize(true)

        orderDateTv.text = getString(R.string.string_format_order_date, Date(order.orderDate).toLongDateFormat())
        subtotalTv.text = getString(R.string.string_format_price, order.subtotal)
        nameTv.text = order.contactInfo.name
        emailTv.text = order.contactInfo.email
        phoneTv.text = order.contactInfo.phone

        orderStatusTv.setText(order.orderStatus.titleRes)
        contactUsButton.setOnClickListener { sendEmailIntent(order.id) }
    }

    private fun sendEmailIntent(orderId: String) {
        val intent = Intent(Intent.ACTION_SENDTO)
        intent.data = Uri.parse("mailto:")
        intent.putExtra(Intent.EXTRA_EMAIL, arrayOf("support@pencake.com"))
        intent.putExtra(Intent.EXTRA_SUBJECT, "Order Support ($orderId)")
        if (intent.resolveActivity(packageManager) != null) {
            startActivity(intent)
        }
    }

    private fun initViewModel() {
        viewModel = withViewModel({ OrderDetailViewModel(dataManager) }) {
            observe(storeInfoLiveData, ::updateStoreInfo)
        }
    }

    private fun updateStoreInfo(storeInfo: StoreInfo?) {
        storeInfo?.let {
            storeNameTv.text = it.name
            storeAddressTv.text = it.address
            storePhoneNumberTv.text = it.phone
            directionsIcon.setOnClickListener { _ ->
                val coordinate = it.coordinate
                val uri = Uri.parse("google.navigation:q=${coordinate.latitude},${coordinate.longitude}")
                val mapIntent = Intent(Intent.ACTION_VIEW, uri)
                mapIntent.`package` = "com.google.android.apps.maps"
                if (mapIntent.resolveActivity(packageManager) != null) {
                    startActivity(mapIntent)
                }
            }
            val mapFragment = fragmentManager.findFragmentById(R.id.mapFragment) as MapFragment
            mapFragment.getMapAsync(this)
        }
    }

    companion object {

        const val EXTRA_ORDER_JSON = "EXTRA_ORDER_JSON"
    }
}